pandas/core/reshape/pivot.py

```diff
@@ -345,12 +329,20 @@ def _all_key(key):
                 table_pieces.append(piece)
                 margin_keys.append(all_key)
         else:
-            margin = grand_margin
             cat_axis = 0
             for key, piece in table.groupby(level=0, axis=cat_axis, observed=observed):
-                all_key = _all_key(key)
+                if len(cols) > 1:
+                    all_key = _all_key(key)
+                else:
+                    all_key = margins_name
                 table_pieces.append(piece)
-                table_pieces.append(Series(margin[key], index=[all_key]))
+                # GH31016 this is to calculate margin for each group, and assign
+                # corresponded key as index
+                transformed_piece = DataFrame(piece.apply(aggfunc)).T
```

Comment:
Does this have any performance impacts?

Reply:
might be? i was thinking of it as well

how to measure it? any suggestions? would like to test it out!

Reply:
We have asvs in benchmarks/reshape.py that would be good to run here at least

Reply:
thanks! @WillAyd i am not very familiar with asv bench, but i see some tests starting with `time_pivot****`, shall I add a new test in there?

Reply:
Run the existing ones first:

```sh
cd asv_bench
asv continuous upstream/master HEAD -b reshape.PivotTable
```

Reply:
![Screen Shot 2020-01-16 at 10 58 20 PM](https://user-images.githubusercontent.com/9269816/72566218-b8dcc900-38b3-11ea-8e33-2e66e8cd39b6.png)

just copied and pasted and tested a bit

Reply:
> Run the existing ones first:
> 
> ```shell
> cd asv_bench
> asv continuous upstream/master HEAD -b reshape.PivotTable
> ```

oops! thanks for the tip!! did not see it 😅 will run

Reply:
@WillAyd i got some issues running asv with python3.7 in virtualenv 😞  so i will investigate a bit tomorrow then, if you already know how to deal with it, would be much appreciated!

Reply:
running asv on current i get: `BENCHMARKS NOT SIGNIFICANTLY CHANGED.`

but i do find an issue with self-review, will investigate and fix tomorrow. thanks for the help on asv @WillAyd